src/core/state.go

```diff
@@ -270,135 +234,94 @@ func (state *BuildState) addActiveTargets(n int) {
 	atomic.AddInt64(&state.progress.numActive, int64(n))
 }
 
-// AddPendingParse adds a task for a pending parse of a build label.
-func (state *BuildState) AddPendingParse(label, dependent BuildLabel, forSubinclude bool) {
+// addPendingParse adds a task for a pending parse of a build label.
+func (state *BuildState) addPendingParse(label, dependent BuildLabel, forSubinclude bool) {
 	atomic.AddInt64(&state.progress.numActive, 1)
 	atomic.AddInt64(&state.progress.numPending, 1)
-	if forSubinclude {
-		state.pendingTasks.Put(pendingTask{Label: label, Dependent: dependent, Type: SubincludeParse})
-	} else {
-		state.pendingTasks.Put(pendingTask{Label: label, Dependent: dependent, Type: Parse})
-	}
+	go func() {
+		defer func() {
+			recover() // Prevent death on 'send on closed channel'
+		}()
+		state.pendingParses <- ParseTask{Label: label, Dependent: dependent, ForSubinclude: forSubinclude}
+	}()
 }
 
 // addPendingBuild adds a task for a pending build of a target.
-func (state *BuildState) addPendingBuild(label BuildLabel, forSubinclude bool) {
-	if forSubinclude {
-		state.addPending(label, SubincludeBuild)
-	} else {
-		state.addPending(label, Build)
-	}
+func (state *BuildState) addPendingBuild(target *BuildTarget) {
+	atomic.AddInt64(&state.progress.numPending, 1)
```

Comment:
How did we get away with not having this here before?

Reply:
It got called within `addPendingTask`, but I ended up deleting that because it's not generic any more. We'll still be calling it the same number of times.